### PR TITLE
Support listing/getting/creating/updating tables/datasets/jobs in other projects

### LINKED ISSUE #1123: BigQuery listTables(DatasetId) does not use project overrided in DatasetId
My Google Application Default Credentials are set to the swast-scratch project, but I wish to list the tables in the `publicdata` project's `samples` dataset. This is a public dataset.

```shell
$ bq ls publicdata:samples
      tableId       Type
 ----------------- -------
  github_nested     TABLE
  github_timeline   TABLE
  gsod              TABLE
  natality          TABLE
  shakespeare       TABLE
  trigrams          TABLE
  wikipedia         TABLE
```

Example code that displays this issue:

```
import com.google.cloud.bigquery.BigQuery;
import com.google.cloud.bigquery.BigQueryOptions;
import com.google.cloud.bigquery.DatasetId;
import com.google.cloud.bigquery.Table;
import java.io.PrintWriter;
import java.util.Iterator;

public class TableLister {
  public static void listTables(PrintWriter out) {
    BigQuery bigquery = BigQueryOptions.defaultInstance().service();
    Iterator<Table> tables = bigquery.listTables(DatasetId.of("publicdata", "samples")).iterateAll();
    while (tables.hasNext()) {
      out.println(tables.next());
    }
  }

  public static void main(String[] args) {
    listTables(new PrintWriter(System.out));
  }
}
```

It fails with the error: `"message" : "Not found: Dataset swast-scratch:samples"`, so it appears that it is not using the project that I set in the DatasetId.

http://googlecloudplatform.github.io/gcloud-java/0.2.5/apidocs/com/google/cloud/bigquery/DatasetId.html#of-java.lang.String-java.lang.String-

### PR DESCRIPTION
**Pull request by mziccard**
This fixes #1123 

@tswast feel free to have a look and comment whether this meets your expectations.

/cc @lesv

### CODE REVIEW

**Comment by lesv** [comment]
LGTM

File: google-cloud-bigquery/src/main/java/com/google/cloud/bigquery/BigQueryImpl.java [hunk 1]
```diff
@@ -265,13 +274,15 @@ public boolean delete(String datasetId, DatasetDeleteOption... options) {
   }
 
   @Override
-  public boolean delete(final DatasetId datasetId, DatasetDeleteOption... options) {
+  public boolean delete(DatasetId datasetId, DatasetDeleteOption... options) {
+    final DatasetId completeDatasetId = datasetId.setProjectId(options().projectId());
```
Comment (tswast) [inline google-cloud-bigquery/src/main/java/com/google/cloud/bigquery/BigQueryImpl.java]: I could see this as possibly causing the same confusion if someone populates the projectId of the DatasetId. (It's certainly less likely that someone would try it, though.)

Maybe raise an error if the projectId is already populated and set to something that doesn't match options().projectId()?

Comment (mziccard) [inline google-cloud-bigquery/src/main/java/com/google/cloud/bigquery/BigQueryImpl.java]: `setProjectId` does not override the value if it's already set, it just set if the user did not provide one (is `null`).

Comment (tswast) [inline google-cloud-bigquery/src/main/java/com/google/cloud/bigquery/BigQueryImpl.java]: Ah, cool, thanks.

**Comment by tswast** [comment]
LGTM

**Pull request merged**